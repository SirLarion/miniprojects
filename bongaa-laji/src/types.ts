export enum MenuState {
    Initial,
    PlayerSelection
}

export enum SpeciesType {
    Foreign = "foreign",
    Native = "native"
}

export type WordSet = {
    singular: string;
    plural: string;
}

export type Card = {
    sIndex: number;
    sType: SpeciesType;
}

export type CardState = {
    selected: boolean;
    revealed: boolean;
    removed: boolean;
}

export type Player = {
    name: string;
    cards: Card[];
}

export type Session = {
    started: boolean;
    numPlayers: number;
    players: Player[];
}
