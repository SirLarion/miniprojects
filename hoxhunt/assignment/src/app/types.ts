export type Stats = {
    agility: number;
    healthpoints: number;
    intelligence: number;
    mana: number;
    resistance: string;
    speed: number;
    stamina: number;
    strength: number;
    weakness: string;
}

export type Skill = {
    name: string;
    element: string;
    damage: number;
}

export type HeroType = {
    name: string;
    imgUrl: string;
    description: string;
    backStory: string;
    attributes: Stats;
    skills: Skill[];
}
