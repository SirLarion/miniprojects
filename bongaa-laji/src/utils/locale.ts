import { WordSet } from '../types';

export const gameName = 'Bongaa Vieraslaji!';

export const startText = 'Aloita peli';

export const playerWords: WordSet = {
    singular: 'Pelaaja',
    plural: 'Pelaajaa'
};

export const cardWords: WordSet = {
    singular: 'Kortti',
    plural: 'Korttia'
};
