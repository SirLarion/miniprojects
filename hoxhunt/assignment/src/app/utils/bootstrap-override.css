/* Override carousel controls' size due to overlap with HeroCard */
.carousel-control-prev { width: 10% !important; }
.carousel-control-next { width: 10% !important; }
