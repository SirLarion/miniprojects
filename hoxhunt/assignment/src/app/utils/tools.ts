export const capitalized = (str: string) => str.replace(/^\w/, (c) => c.toUpperCase());
