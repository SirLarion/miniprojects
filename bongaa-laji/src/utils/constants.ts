
import { Session, Player, Card, SpeciesType } from '../types';
import { playerWords } from '../utils/locale';

export const cardH = 14;
export const cardW = 10;

export const minPlayers = 2;
export const maxPlayers = 4;
export const speciesAmount = 24;

const rowAmount = 3;
const colAmount = 8;

export const gridW = speciesAmount/rowAmount * cardW + speciesAmount/rowAmount -1;
export const gridH = speciesAmount/colAmount * cardH + speciesAmount/colAmount -1;

export const defaultCards: Card[] = (new Array(speciesAmount)).fill(0).map((n: number, i: number) => {
    const c: Card = {
        sIndex: (i % 12) + 1,
        sType: i < 12 ? SpeciesType.Foreign : SpeciesType.Native
    }
    return c;
});

export const defaultPlayers: Player[] = [
    {
        name: `${playerWords.singular} 1`,
        cards: []  
    },
    {
        name: `${playerWords.singular} 2`,
        cards: []  
    }
];

export const defaultSession: Session = {
    started: false,
    numPlayers: minPlayers,
    players: defaultPlayers
}


