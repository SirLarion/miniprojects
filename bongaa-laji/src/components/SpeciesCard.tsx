import styled from 'styled-components';

import { FlexContainer } from './Layout';
import { Card, CardState } from '../types';

import { cardH, cardW } from '../utils/constants';

const H = cardH;
const W = cardW;

const Base = styled.div`
    width: ${W}rem;
    height: ${H}rem;
    position: relative;
    background-color: ${props => props.theme.colors.mainDark};
    border-radius: 1rem;
`;

interface ICard {
    selected: boolean;
    hidden: boolean;
}
const CardOuter = styled.div`
    ${(p: ICard) => (
        p.selected 
            ? `z-index: 20; 
               width: ${2*W}rem; 
               height: ${2*H}rem; 
               top: 29vh; left: 41.66vw;
               padding: 1rem !important;
               box-shadow: 0px 0px 2rem rgba(0, 0, 0, 0.20);
               `
            : `width: ${W}rem; height: ${H}rem;`
    )}
    ${(p: ICard) => p.hidden? 'display: none;' : ''}
    background-color: ${props => props.theme.colors.secondary};
    position: fixed;
    padding: 0.5rem;
    border-radius: 1rem;
    align-items: flex-start;
`;

const CardInner = styled(FlexContainer)`
    width: 100%;
    height: 100%;
    background-color: white;
    border-radius: 0.5rem;
    border: 0.125rem solid #CBCBCB;
    justify-content: center; 
    align-items: center;
`;

const CardImage = styled.img`
    width: inherit;
    border-style: none;
`;

interface SpeciesProps {
    boardIndex: number;
    handleSelect: (event: React.MouseEvent) => void;
    state: CardState;
    card: Card;
}

const SpeciesCard: React.FC<SpeciesProps> = ({boardIndex, handleSelect, state, card}) => {
    return (
        <Base>
            <CardOuter 
                onClick={handleSelect} 
                selected={state.selected}
                hidden={state.removed && !state.revealed}
            >
                <CardInner>
                    <CardImage 
                        src={`/${card.sType}/species${card.sIndex}_${state.revealed ? 'back' : 'front'}.jpg`} 
                        alt="species" 
                    />
                </CardInner>
            </CardOuter>
        </Base>
    );
}

export default SpeciesCard;
