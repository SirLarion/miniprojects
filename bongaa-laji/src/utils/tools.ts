import seedrandom from 'seedrandom';

import { Card, Player } from '../types'
import { cardWords } from '../utils/locale';

// Shuffle the given array of Card elements randomly or with a
// seed if one is given
export const shuffleCards = (a: Card[], seed?: string) => {
    var j, x, i;
    var b = a.slice();
    const s = seed ? seed : `${Math.random()}`;
    seedrandom(s, { global: true });

    for (i = b.length - 1; i > 0; i--) {
        j = Math.floor(Math.random() * (i + 1));
        x = b[i];
        b[i] = b[j];
        b[j] = x;
    }
    return b;
}

// Count the amount of true values in an array of booleans 
// where each value corresponds to a card with a foreign species
// that has been removed from the board
export const countForeignRemaining = (a: boolean[]): number => {
    var count = 0;
    for(var i = 0; i < a.length; ++i){
        if(a[i])
            count++;
    }
    return a.length/2 - count;
}

// Filter the players with the most cards from the list of players
// given as a parameter
const getWinners = (players: Player[]): Player[] => {

    let winners: Player[] = [];
    const playerCards = players.map(p => p.cards.length);
    const max = Math.max(...playerCards);

    for(var i = 0; i < players.length; i++){
        if(playerCards[i] === max)
            winners.push(players[i]);
    }
    return winners;
}

export const getWinnerText = (players: Player[]): string[] => {
    const numPlayers = players.length;
    const winners = getWinners(players);
    const winnerNames = winners.map(w => w.name);
    const maxCards = winners[0].cards.length;

    let text = '';
    let subtext = '';

    if(winners.length === 1) {
        text = `${winners[0].name} voittaa!`;
        subtext = `${maxCards} ${maxCards === 1 ? cardWords.singular : cardWords.plural}`;
    }
    else if(winners.length > 1) {
        if(winners.length === numPlayers)
            text = 'Tasapeli!';
        else
            text = `${winnerNames.slice(0, -1).join(', ')} ja ${winnerNames[winners.length -1]} voittavat!`;

        subtext = `${maxCards} ${maxCards === 1 ? cardWords.singular : cardWords.plural}`;
    }
    return [text, subtext];
}
