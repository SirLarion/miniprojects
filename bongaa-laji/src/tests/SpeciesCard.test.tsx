import { render, screen, fireEvent, act } from '@testing-library/react';

import SpeciesCard from '../components/SpeciesCard';
import { Card, SpeciesType } from '../types';
import * as Tools from '../utils/tools';

describe('<SpeciesCard>', () => {
    //-----------------------------------------------------------------------//
    it('should match the standard snapshot', () => {
        const testCard: Card = { sIndex: 3, sType: SpeciesType.Native }
        const { container } = render(
            <SpeciesCard  
                boardIndex={3}
                handleSelect={jest.fn()}
                state={{selected: false, revealed: false}}
                card={testCard}
            />
        );
        expect(container).toMatchSnapshot();
    });

    //-----------------------------------------------------------------------//
    it('should contain a visible image by default', () => {
        const testCard: Card = { sIndex: 3, sType: SpeciesType.Native }
        const { getByAltText } = render(
            <SpeciesCard  
                boardIndex={3}
                handleSelect={jest.fn()}
                state={{selected: false, revealed: false}}
                card={testCard}
            />
        );
        expect(getByAltText('species')).toBeVisible();
    });

    //-----------------------------------------------------------------------//
    it('should call handleSelect when clicked', () => {
        const mockClick = jest.fn();
        const testCard: Card = { sIndex: 3, sType: SpeciesType.Native }
        const { container } = render(
            <SpeciesCard  
                boardIndex={3}
                handleSelect={mockClick}
                state={{selected: false, revealed: false}}
                card={testCard}
            />
        );
        fireEvent.click(container.firstChild.firstChild);
        expect(mockClick.mock.calls.length).toBe(1);
    });


});
