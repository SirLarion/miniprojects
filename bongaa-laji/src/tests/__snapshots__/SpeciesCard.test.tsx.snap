// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpeciesCard> should match the standard snapshot 1`] = `
<div>
  <div
    class="sc-jSgupP kwgFEy"
  >
    <div
      class="sc-gKsewC llIhqD"
    >
      <div
        class="sc-hKgILt sc-iBPRYJ Axa-Dd YMIft"
      >
        <img
          alt="species"
          class="sc-fubCfw cHUIOF"
          src="/native/species3_front.jpg"
        />
      </div>
    </div>
  </div>
</div>
`;
